### PR TITLE
Pretrain multipack

### PR DESCRIPTION
**Pull request by winglian**
multipack for pretraining had some wonky edge cases that were broken and untested

### CODE REVIEW

File: src/axolotl/core/trainer_builder.py [hunk 1]
```diff
@@ -1879,6 +1879,8 @@ def build_collator(
         if training_args.pretraining:
             if self.cfg.pretraining_sample_concatenation is False:
                 return DataCollatorForSeq2Seq(self.tokenizer, **kwargs)
+            if self.cfg.micro_batch_size > 1:
+                return DataCollatorForSeq2Seq(self.tokenizer, **kwargs)
```
Comment (NanoCode012) [inline src/axolotl/core/trainer_builder.py:1884]: Are we expecting this to be the original `micro_batch_size` or would it have been overwritten in `wrap_pretraining_dataset` to be set to 1?

File: src/axolotl/utils/data/pretraining.py [hunk 2]
```diff
@@ -248,8 +246,6 @@ def encode_packed_pretraining(
     max_seq_length: int = 2048,
     batch_size: int = 4,
     multipack_attn: Optional[bool] = False,
```
Comment (NanoCode012) [inline src/axolotl/utils/data/pretraining.py]: Should `multipack_attn` default to True following Pydantic model?

File: tests/e2e/test_llama_pretrain.py [hunk 3]
```diff
@@ -68,3 +75,12 @@ def test_pretrain(self, temp_dir, sample_packing):
 
         train(cfg=cfg, dataset_meta=dataset_meta)
         check_model_output_exists(temp_dir, cfg)
+        loss_threshold = 3.5
+        if sample_packing and not pretrain_multipack_attn:
+            loss_threshold = 6.5
```
Comment (NanoCode012) [inline tests/e2e/test_llama_pretrain.py:80]: Why would the threshold be this high without multipack_attn?

File: tests/test_packed_pretraining.py [hunk 4]
```diff
@@ -87,9 +88,9 @@ def test_packing_stream_dataset(self):
             assert data["labels"].shape == torch.Size(
                 [1, original_bsz * cfg.sequence_len]
             )
-            assert data["attention_mask"].shape == torch.Size(
-                [1, original_bsz * cfg.sequence_len]
-            )
+            # assert data["attention_mask"].shape == torch.Size(
```
Comment (NanoCode012) [inline tests/test_packed_pretraining.py:93]: Perhaps , we can add `assert "attention_mask" not in data`

File: tests/e2e/test_llama_pretrain.py (hunk 3 above)
Comment (winglian) [reply tests/e2e/test_llama_pretrain.py:80]: it's the same as concatenation so their is cross-sample contamination resulting in a higher loss
